proxy/shadow_test.go

```diff
@@ -0,0 +1,398 @@
+		routes     string
+		timeout    time.Duration
+		mainFunc   func()
+		shadowFunc func()
+		check      func(va *httptest.VegetaAttacker) error
+		debug      bool
+	}{
+		{
+			name: "50%",
+			routes: `
+main: PathSubtree("/")
+  -> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+  -> teeLoopback("test")
+  -> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+  -> "%s";
+	`,
+			debug: true,
+		},
+		{
+			name: "50% with fifo",
+			routes: `
+main: PathSubtree("/")
+  -> fifo(10, 5, "1s")
+  -> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+  -> fifo(10, 5, "1s")
+  -> teeLoopback("test")
+  -> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+  -> fifo(10, 5, "1s")
+  -> "%s";
+	`,
+			debug: true,
+		},
+		{
+			name: "50% with fifo and slow shadow",
+			routes: `
+main: PathSubtree("/")
+  -> fifo(10, 5, "1s")
+  -> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+  -> fifo(10, 5, "1s")
+  -> teeLoopback("test")
+  -> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+  -> fifo(10, 5, "1s")
+  -> "%s";
+	`,
+			shadowFunc: func() { time.Sleep(500 * time.Millisecond) },
+			debug:      true,
+		},
+		{
+			name: "50% with fifo and shadow times out",
+			routes: `
+main: PathSubtree("/")
+  -> fifo(10, 5, "1s")
+  -> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+  -> fifo(10, 5, "1s")
+  -> teeLoopback("test")
+  -> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+  -> fifo(10, 5, "1s")
+  -> "%s";
+	`,
+			shadowFunc: func() { time.Sleep(1100 * time.Millisecond) },
+			debug:      true,
+		},
+		{
+			name: "50% with fifo and slow main",
+			routes: `
+main: PathSubtree("/")
+  -> fifo(10, 5, "1s")
+  -> "%s";
+split: PathSubtree("/") && Traffic(0.5)
+  -> fifo(10, 5, "1s")
+  -> teeLoopback("test")
+  -> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+  -> fifo(10, 5, "1s")
+  -> "%s";
+	`,
+			mainFunc: func() { time.Sleep(50 * time.Millisecond) },
+			debug:    true,
+		},
+		{
+			name: "100% shadow with fifo and 25% timing out main",
+			routes: `
+main: PathSubtree("/")
+  -> fifo(2, 5, "200ms")
+  -> "%s";
+split: PathSubtree("/") && Traffic(1.0)
+  -> fifo(2, 5, "200ms")
+  -> teeLoopback("test")
+  -> "%s";
+shadow: PathSubtree("/") && Tee("test") && True()
+  -> fifo(2, 5, "200ms")
+  -> "%s";
+	`,
+			mainFunc: func() {
+				if rand.Float64() < 0.25 {
+					time.Sleep(250 * time.Millisecond)
+				}
+			},
+			debug: true,
+			check: func(va *httptest.VegetaAttacker) error {
+				statusFifoFull, _ := va.CountStatus(http.StatusServiceUnavailable)
+				statusFifoTimeout, _ := va.CountStatus(http.StatusBadGateway)
+				statusFifoErr, _ := va.CountStatus(http.StatusInternalServerError)
+
+				if statusFifoFull == 0 {
+					return fmt.Errorf("fifo full %d", statusFifoFull)
+				}
+				if statusFifoTimeout == 0 {
+					return fmt.Errorf("fifo timeout %d", statusFifoTimeout)
+				}
+				if statusFifoErr != 0 {
+					return fmt.Errorf("fifo err %d", statusFifoErr)
+				}
+				return nil
+			},
```

Comment:
This is quite hard to read when placed in the table test. Alternatively you could have a test helper method that setups proxy and runs load and then use it in a series of nested t.Runs like
```go
t.Run("foo", func(t *testing.T) {
	va := runTest(`
	  main: PathSubtree("/")
            -> fifo(2, 5, "200ms")
            -> "$backendURL";
          split: PathSubtree("/") && Traffic(1.0)
            -> fifo(2, 5, "200ms")
            -> teeLoopback("test")
            -> "$backendURL";
          shadow: PathSubtree("/") && Tee("test") && True()
            -> fifo(2, 5, "200ms")
            -> "$shadowURL";
	`)
	...
})
```